docs/architecture/adr-065-store-v2.md

```diff
@@ -0,0 +1,195 @@
+
+* Since each module has its own IAVL `KVStore`, commitments are not [atomic](https://github.com/cosmos/cosmos-sdk/issues/14625)
+    * Note, we can still allow modules to have their own IAVL `KVStore`, but the
+      IAVL library will need to support the ability to pass a DB instance as an
+      argument to various IAVL APIs.
+* Since IAVL is responsible for both state storage and commitment, running an 
+  archive node becomes increasingly expensive as disk space grows exponentially.
+* As the size of a network increases, various performance bottlenecks start to
+  emerge in many areas such as query performance, network upgrades, state
+  migrations, and general application performance.
+* Developer UX is poor as it does not allow application developers to experiment
+  with different types of approaches to storage and commitments, along with the
+  complications of many layers of abstractions referenced above.
+
+See the [Storage Discussion](https://github.com/cosmos/cosmos-sdk/discussions/13545) for more information.
+
+## Alternatives
+
+There was a previous attempt to refactor the storage layer described in [ADR-040](./adr-040-storage-and-smt-state-commitments.md).
+However, this approach mainly stems on the short comings of IAVL and various performance
+issues around it. While there was a (partial) implementation of [ADR-040](./adr-040-storage-and-smt-state-commitments.md),
+it was never adopted for a variety of reasons, such as the reliance on using an
+SMT, which was more in a research phase, and some design choices that couldn't
+be fully agreed upon, such as the snap-shotting mechanism that would result in
+massive state bloat.
+
+## Decision
+
+We propose to build upon some of the great ideas introduced in [ADR-040](./adr-040-storage-and-smt-state-commitments.md),
+while being a bit more flexible with the underlying implementations and overall
+less intrusive. Specifically, we propose to:
+
+* Separate the concerns of state commitment (**SC**), needed for consensus, and
+  state storage (**SS**), needed for state machine and clients.
+* Reduce layers of abstractions necessary between the RMS and underlying stores.
+* Provide atomic module store commitments by providing a batch database object
+  to core IAVL APIs.
+* Reduce complexities in the `CacheKVStore` implementation while also improving
+  performance<sup>[3]</sup>.
+
+Furthermore, we will keep the IAVL is the backing [commitment](https://cryptography.fandom.com/wiki/Commitment_scheme)
+store for the time being. While we might not fully settle on the use of IAVL in
+the long term, we do not have strong empirical evidence to suggest a better
+alternative. Given that the SDK provides interfaces for stores, it should be sufficient
+to change the backing commitment store in the future should evidence arise to
+warrant a better alternative. However there is promising work being done to IAVL
+that should result in significant performance improvement <sup>[1,2]</sup>.
+
```

Comment:
one thing to note with this refactor is we should think about removing the fast node system in iavl because it was a poor mans SS & SC separation. With this redesign we can remove it